Update ConstConfigIOGroup config file schema

avilcheslopez opened the pull request:
Update schema to allow specifying a default value for a signal. For each signal, either a default value or an array of values needs to be provided.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

- Relates to #2731 
- Fixes #2732

avilcheslopez committed:
Update ConstConfigIOGroup config file schema

Update schema to allow specifying a default value for a signal. For each
signal, either a default value or an array of values needs to be provided.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Enable ConstConfigIOGroup to handle default values

Enable ConstConfigIOGroup to handle either a default value or an array
of values for each signal (according to updated schema).

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -88,6 +88,7 @@ namespace geopm
                 int domain;
                 std::function<double(const std::vector<double> &)> agg_function;
                 std::string description;
+                bool default_value;
```
Nit on naming.  default_value is a boolean indicating if a devalue value exists but might be interpreted as the actual default value.  default_value_provided is used elsewhere and seems somewhat clearer...or maybe default_value_exists or default_value_present?  

Otherwise lgtm.

avilcheslopez committed:
Update ConstConfigIOGroup unit tests

- Update unit tests after changes to ConstConfigIOGroup config file
schema
- Add unit tests for handling a single default value for signals

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Rename "default_value" in ConstConfigIOGroup schema

- Rename "default_value" field to "common_value"
- Adjust/update tests
- Minor clean-up/improvements

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -203,13 +209,21 @@ namespace geopm
         }
 
-        return it->second->values[domain_idx];
+        double value = 0.0;
```
Other IOGroups use NAN as a default value as a precaution; if we get to the return without setting the value the NAN indicates something has gone wrong.  Should we follow the same pattern here?

avilcheslopez replied:
Yeah, that makes sense. I'll make that change.

avilcheslopez committed:
In ConstConfigIOGroup, init values with NAN by default

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -88,6 +88,7 @@ namespace geopm
                 int domain;
                 std::function<double(const std::vector<double> &)> agg_function;
                 std::string description;
+                bool common_value_provided;
```
`common_valud_provided` -> `is_common_value` or `is _common_value_provided`

I personally prefer `is_common_value`, but i'm okay with the longer name.  I try to enforce prefixing bool variable names with short verbs (usually a "to be verb") to aid with reading logic statements in code.  There are a couple of similar stack variables in the parser.

cmcantalupo reviewed (commented):
Minor nit on struct name, otherwise looks great.

avilcheslopez committed:
Minor clean-up in ConstConfigIOGroup

Rename member "common_value_provided" to "is_common_value_provided"
for clarity.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>
